src/main/java/org/opensearch/knn/index/codec/nativeindex/NativeIndexWriter.java

```diff
@@ -0,0 +1,287 @@
+    }
+
+    /**
+     * flushes the index
+     * @param knnVectorValues
+     * @return boolean to indicate if the index was written
+     * @throws IOException
+     */
+    public boolean flushIndex(final KNNVectorValues<?> knnVectorValues) throws IOException {
+        if (knnVectorValues.docId() == -1) {
+            knnVectorValues.nextDoc();
+            knnVectorValues.getVector();
+        }
+
+        recordRefreshStats();
+        return buildAndWriteIndex(knnVectorValues);
+    }
+
+    /**
+     * Merges kNN index
+     * @param knnVectorValues
+     * @return boolean to indicate if the index was written
+     * @throws IOException
+     */
+    public boolean mergeIndex(final KNNVectorValues<?> knnVectorValues) throws IOException {
+        if (knnVectorValues.docId() == -1) {
+            knnVectorValues.nextDoc();
+            knnVectorValues.getVector();
+        }
+
+        long arraySize = knnVectorValues.bytesPerVector();
+        startMergeStats(knnVectorValues.dimension(), arraySize);
+        boolean indexWritten = buildAndWriteIndex(knnVectorValues);
+        endMergeStats(knnVectorValues.dimension(), arraySize);
+
+        return indexWritten;
+    }
+
+    // The logic for building parameters need to be cleaned up. There are various cases handled here
+    // Currently it falls under two categories - with model and without model. Without model is further divided based on vector data type
+    // TODO: Refactor this so its scalable. Possibly move it out of this class
+    private BuildIndexParams indexParams(FieldInfo fieldInfo, String indexPath, KNNEngine knnEngine) throws IOException {
+        final Map<String, Object> parameters;
+        final VectorDataType vectorDataType;
+        if (fieldInfo.attributes().containsKey(MODEL_ID)) {
+            Model model = getModel(fieldInfo);
+            vectorDataType = model.getModelMetadata().getVectorDataType();
+            parameters = getTemplateParameters(fieldInfo, model);
+        } else {
+            vectorDataType = VectorDataType.get(
+                fieldInfo.attributes().getOrDefault(KNNConstants.VECTOR_DATA_TYPE_FIELD, VectorDataType.DEFAULT.getValue())
+            );
+            parameters = getParameters(fieldInfo, vectorDataType, knnEngine);
+        }
+
+        return BuildIndexParams.builder()
+            .parameters(parameters)
+            .vectorDataType(vectorDataType)
+            .knnEngine(knnEngine)
+            .indexPath(indexPath)
+            .build();
+    }
+
+    private Map<String, Object> getParameters(FieldInfo fieldInfo, VectorDataType vectorDataType, KNNEngine knnEngine) throws IOException {
+        Map<String, Object> parameters = new HashMap<>();
+        Map<String, String> fieldAttributes = fieldInfo.attributes();
+        String parametersString = fieldAttributes.get(KNNConstants.PARAMETERS);
+
+        // parametersString will be null when legacy mapper is used
+        if (parametersString == null) {
```

Comment:
Can we simply this to make more understandable 

if (Parameter == null) then buildLegacyParms() else put that para string in paramter 
next  updateFaissBinaryParameters
next update indexing thread

Reply:
The logic for this is kept as it is for now. This surely needs a refactor but considering the amount of changes going in this PR I just added a comment above  `indexParams` I can follow up after the PR in terms of cleaning this part up